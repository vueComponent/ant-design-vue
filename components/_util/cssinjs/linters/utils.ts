import devWarning from '../../../vc-util/warning';
import type { LinterInfo } from './interface';

export function lintWarning(message: string, info: LinterInfo) {
  const { path, parentSelectors } = info;

  devWarning(
    false,
    `[Ant Design Vue CSS-in-JS] ${path ? `Error in '${path}': ` : ''}${message}${
      parentSelectors.length ? ` Selector info: ${parentSelectors.join(' -> ')}` : ''
    }`,
  );
}
