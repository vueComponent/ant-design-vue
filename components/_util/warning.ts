import vcWarning, { resetWarned } from '../vc-util/warning';

export { resetWarned };
export function noop() {}

type Warning = (valid: boolean, component: string, message?: string) => void;

// eslint-disable-next-line import/no-mutable-exports
let warning: Warning = noop;
if (process.env.NODE_ENV !== 'production') {
  warning = (valid, component, message) => {
    vcWarning(valid, `[ant-design-vue: ${component}] ${message}`);

    // StrictMode will inject console which will not throw warning in React 17.
    if (process.env.NODE_ENV === 'test') {
      resetWarned();
    }
  };
}

export default warning;
