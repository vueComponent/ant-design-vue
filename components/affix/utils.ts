import addEventListener from '../vc-util/Dom/addEventListener';
import supportsPassive from '../_util/supportsPassive';

export type BindElement = HTMLElement | Window | null | undefined;

export function getTargetRect(target: BindElement): DOMRect {
  return target !== window
    ? (target as HTMLElement).getBoundingClientRect()
    : ({ top: 0, bottom: window.innerHeight } as DOMRect);
}

export function getFixedTop(placeholderRect: DOMRect, targetRect: DOMRect, offsetTop: number) {
  if (offsetTop !== undefined && targetRect.top > placeholderRect.top - offsetTop) {
    return `${offsetTop + targetRect.top}px`;
  }
  return undefined;
}

export function getFixedBottom(
  placeholderRect: DOMRect,
  targetRect: DOMRect,
  offsetBottom: number,
) {
  if (offsetBottom !== undefined && targetRect.bottom < placeholderRect.bottom + offsetBottom) {
    const targetBottomOffset = window.innerHeight - targetRect.bottom;
    return `${offsetBottom + targetBottomOffset}px`;
  }
  return undefined;
}

// ======================== Observer ========================
const TRIGGER_EVENTS = [
  'resize',
  'scroll',
  'touchstart',
  'touchmove',
  'touchend',
  'pageshow',
  'load',
];

interface ObserverEntity {
  target: HTMLElement | Window;
  affixList: any[];
  eventHandlers: { [eventName: string]: any };
}

let observerEntities: ObserverEntity[] = [];

export function getObserverEntities() {
  // Only used in test env. Can be removed if refactor.
  return observerEntities;
}

export function addObserveTarget<T>(target: HTMLElement | Window | null, affix: T): void {
  if (!target) return;

  let entity = observerEntities.find(item => item.target === target);

  if (entity) {
    entity.affixList.push(affix);
  } else {
    entity = {
      target,
      affixList: [affix],
      eventHandlers: {},
    };
    observerEntities.push(entity);

    // Add listener
    TRIGGER_EVENTS.forEach(eventName => {
      entity!.eventHandlers[eventName] = addEventListener(target, eventName, () => {
        entity!.affixList.forEach(
          targetAffix => {
            const { lazyUpdatePosition } = (targetAffix as any).exposed;
            lazyUpdatePosition();
          },
          (eventName === 'touchstart' || eventName === 'touchmove') && supportsPassive
            ? ({ passive: true } as EventListenerOptions)
            : false,
        );
      });
    });
  }
}

export function removeObserveTarget<T>(affix: T): void {
  const observerEntity = observerEntities.find(oriObserverEntity => {
    const hasAffix = oriObserverEntity.affixList.some(item => item === affix);
    if (hasAffix) {
      oriObserverEntity.affixList = oriObserverEntity.affixList.filter(item => item !== affix);
    }
    return hasAffix;
  });

  if (observerEntity && observerEntity.affixList.length === 0) {
    observerEntities = observerEntities.filter(item => item !== observerEntity);

    // Remove listener
    TRIGGER_EVENTS.forEach(eventName => {
      const handler = observerEntity.eventHandlers[eventName];
      if (handler && handler.remove) {
        handler.remove();
      }
    });
  }
}
