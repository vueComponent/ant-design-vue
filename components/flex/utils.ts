import classNames from '../_util/classNames';

import type { FlexProps } from './interface';

export const flexWrapValues = ['wrap', 'nowrap', 'wrap-reverse'] as const;

export const justifyContentValues = [
  'flex-start',
  'flex-end',
  'start',
  'end',
  'center',
  'space-between',
  'space-around',
  'space-evenly',
  'stretch',
  'normal',
  'left',
  'right',
] as const;

export const alignItemsValues = [
  'center',
  'start',
  'end',
  'flex-start',
  'flex-end',
  'self-start',
  'self-end',
  'baseline',
  'normal',
  'stretch',
] as const;

const genClsWrap = (prefixCls: string, props: FlexProps) => {
  const wrapCls: Record<PropertyKey, boolean> = {};
  flexWrapValues.forEach(cssKey => {
    wrapCls[`${prefixCls}-wrap-${cssKey}`] = props.wrap === cssKey;
  });
  return wrapCls;
};

const genClsAlign = (prefixCls: string, props: FlexProps) => {
  const alignCls: Record<PropertyKey, boolean> = {};
  alignItemsValues.forEach(cssKey => {
    alignCls[`${prefixCls}-align-${cssKey}`] = props.align === cssKey;
  });
  alignCls[`${prefixCls}-align-stretch`] = !props.align && !!props.vertical;
  return alignCls;
};

const genClsJustify = (prefixCls: string, props: FlexProps) => {
  const justifyCls: Record<PropertyKey, boolean> = {};
  justifyContentValues.forEach(cssKey => {
    justifyCls[`${prefixCls}-justify-${cssKey}`] = props.justify === cssKey;
  });
  return justifyCls;
};

function createFlexClassNames(prefixCls: string, props: FlexProps) {
  return classNames({
    ...genClsWrap(prefixCls, props),
    ...genClsAlign(prefixCls, props),
    ...genClsJustify(prefixCls, props),
  });
}

export default createFlexClassNames;
