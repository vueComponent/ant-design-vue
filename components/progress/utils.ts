import { presetPrimaryColors } from '@ant-design/colors';
import devWarning from '../vc-util/devWarning';
import type { CircleProps } from './Circle';
import type { ProgressProps } from './props';

export function validProgress(progress: number | undefined) {
  if (!progress || progress < 0) {
    return 0;
  }
  if (progress > 100) {
    return 100;
  }
  return progress;
}

export function getSuccessPercent({ success, successPercent }: ProgressProps) {
  let percent = successPercent;
  /** @deprecated Use `percent` instead */
  if (success && 'progress' in success) {
    devWarning(
      false,
      'Progress',
      '`success.progress` is deprecated. Please use `success.percent` instead.',
    );
    percent = success.progress;
  }
  if (success && 'percent' in success) {
    percent = success.percent;
  }
  return percent;
}

export function getPercentage({ percent, success, successPercent }: ProgressProps) {
  const realSuccessPercent = validProgress(getSuccessPercent({ success, successPercent }));
  return [realSuccessPercent, validProgress(validProgress(percent) - realSuccessPercent)];
}

export function getStrokeColor({
  success = {},
  strokeColor,
}: Partial<CircleProps>): (string | Record<string, string>)[] {
  const { strokeColor: successColor } = success;
  return [successColor || presetPrimaryColors.green, strokeColor || null!];
}

export const getSize = (
  size: ProgressProps['size'],
  type: ProgressProps['type'] | 'step',
  extra?: {
    steps?: number;
    strokeWidth?: number;
  },
): { width: number; height: number } => {
  let width = -1;
  let height = -1;
  if (type === 'step') {
    const steps = extra!.steps!;
    const strokeWidth = extra!.strokeWidth!;
    if (typeof size === 'string' || typeof size === 'undefined') {
      width = size === 'small' ? 2 : 14;
      height = strokeWidth ?? 8;
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = 14, height = 8] = size;
    }
    width *= steps;
  } else if (type === 'line') {
    const strokeWidth = extra?.strokeWidth;
    if (typeof size === 'string' || typeof size === 'undefined') {
      height = strokeWidth || (size === 'small' ? 6 : 8);
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      [width = -1, height = 8] = size;
    }
  } else if (type === 'circle' || type === 'dashboard') {
    if (typeof size === 'string' || typeof size === 'undefined') {
      [width, height] = size === 'small' ? [60, 60] : [120, 120];
    } else if (typeof size === 'number') {
      [width, height] = [size, size];
    } else {
      if (process.env.NODE_ENV !== 'production') {
        devWarning(
          false,
          'Progress',
          'Type "circle" and "dashboard" do not accept array as `size`, please use number or preset size instead.',
        );
      }

      width = size[0] ?? size[1] ?? 120;
      height = size[0] ?? size[1] ?? 120;
    }
  }
  return { width, height };
};
