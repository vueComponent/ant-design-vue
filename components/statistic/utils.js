import moment from 'moment';
import padStart from 'lodash/padStart';

import interopDefault from '../_util/interopDefault';

// Countdown
const timeUnits = [
  ['Y', 1000 * 60 * 60 * 24 * 365], // years
  ['M', 1000 * 60 * 60 * 24 * 30], // months
  ['D', 1000 * 60 * 60 * 24], // days
  ['H', 1000 * 60 * 60], // hours
  ['m', 1000 * 60], // minutes
  ['s', 1000], // seconds
  ['S', 1], // million seconds
];

export function formatTimeStr(duration, format) {
  let leftDuration = duration;

  const escapeRegex = /\[[^\]]*\]/g;
  const keepList = (format.match(escapeRegex) || []).map(str => str.slice(1, -1));
  const templateText = format.replace(escapeRegex, '[]');

  const replacedText = timeUnits.reduce((current, [name, unit]) => {
    if (current.indexOf(name) !== -1) {
      const value = Math.floor(leftDuration / unit);
      leftDuration -= value * unit;
      return current.replace(new RegExp(`${name}+`, 'g'), match => {
        const len = match.length;
        return padStart(value.toString(), len, '0');
      });
    }
    return current;
  }, templateText);

  let index = 0;
  return replacedText.replace(escapeRegex, () => {
    const match = keepList[index];
    index += 1;
    return match;
  });
}

export function formatCountdown(value, config) {
  const { format = '' } = config;
  const target = interopDefault(moment)(value).valueOf();
  const current = interopDefault(moment)().valueOf();
  const diff = Math.max(target - current, 0);
  return formatTimeStr(diff, format);
}
