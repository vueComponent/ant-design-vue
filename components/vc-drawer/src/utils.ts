export function dataToArray(vars: any) {
  if (Array.isArray(vars)) {
    return vars;
  }
  return [vars];
}
const transitionEndObject: Record<string, string> = {
  transition: 'transitionend',
  WebkitTransition: 'webkitTransitionEnd',
  MozTransition: 'transitionend',
  OTransition: 'oTransitionEnd otransitionend',
};
export const transitionStr: string = Object.keys(transitionEndObject).filter(key => {
  if (typeof document === 'undefined') {
    return false;
  }
  const html = document.getElementsByTagName('html')[0];
  return key in (html ? html.style : {});
})[0];
export const transitionEndFun: string = transitionEndObject[transitionStr];

export function addEventListener(
  target: HTMLElement,
  eventType: string,
  callback: (e: TouchEvent | Event) => void,
  options?: any,
) {
  if (target.addEventListener) {
    target.addEventListener(eventType, callback, options);
  } else if ((target as any).attachEvent) {
    // tslint:disable-line
    (target as any).attachEvent(`on${eventType}`, callback); // tslint:disable-line
  }
}

export function removeEventListener(
  target: HTMLElement,
  eventType: string,
  callback: (e: TouchEvent | Event) => void,
  options?: any,
) {
  if (target.removeEventListener) {
    target.removeEventListener(eventType, callback, options);
  } else if ((target as any).attachEvent) {
    // tslint:disable-line
    (target as any).detachEvent(`on${eventType}`, callback); // tslint:disable-line
  }
}

export function transformArguments(arg: any, cb: any) {
  const result = typeof arg === 'function' ? arg(cb) : arg;
  if (Array.isArray(result)) {
    if (result.length === 2) {
      return result;
    }
    return [result[0], result[1]];
  }
  return [result];
}

export const isNumeric = (value: string | number | undefined) =>
  !isNaN(parseFloat(value as string)) && isFinite(value as number);

export const windowIsUndefined = !(
  typeof window !== 'undefined' &&
  window.document &&
  window.document.createElement
);

export const getTouchParentScroll = (
  root: HTMLElement,
  currentTarget: HTMLElement | Document | null,
  differX: number,
  differY: number,
): boolean => {
  if (!currentTarget || currentTarget === document || currentTarget instanceof Document) {
    return false;
  }
  // root 为 drawer-content 设定了 overflow, 判断为 root 的 parent 时结束滚动；
  if (currentTarget === root.parentNode) {
    return true;
  }

  const isY = Math.max(Math.abs(differX), Math.abs(differY)) === Math.abs(differY);
  const isX = Math.max(Math.abs(differX), Math.abs(differY)) === Math.abs(differX);

  const scrollY = currentTarget.scrollHeight - currentTarget.clientHeight;
  const scrollX = currentTarget.scrollWidth - currentTarget.clientWidth;

  const style = document.defaultView.getComputedStyle(currentTarget);
  const overflowY = style.overflowY === 'auto' || style.overflowY === 'scroll';
  const overflowX = style.overflowX === 'auto' || style.overflowX === 'scroll';

  const y = scrollY && overflowY;
  const x = scrollX && overflowX;

  if (
    (isY &&
      (!y ||
        (y &&
          ((currentTarget.scrollTop >= scrollY && differY < 0) ||
            (currentTarget.scrollTop <= 0 && differY > 0))))) ||
    (isX &&
      (!x ||
        (x &&
          ((currentTarget.scrollLeft >= scrollX && differX < 0) ||
            (currentTarget.scrollLeft <= 0 && differX > 0)))))
  ) {
    return getTouchParentScroll(root, currentTarget.parentNode as HTMLElement, differX, differY);
  }
  return false;
};
