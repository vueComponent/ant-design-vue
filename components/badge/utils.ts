import { PresetColorTypes, PresetStatusColorTypes } from '../_util/colors';

export function isPresetColor(color?: string): boolean {
  return (
    (PresetColorTypes as any[]).indexOf(color) !== -1 ||
    (PresetStatusColorTypes as any[]).indexOf(color) !== -1
  );
}
