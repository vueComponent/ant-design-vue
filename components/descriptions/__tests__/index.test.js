import { mount } from '@vue/test-utils';
import { h } from 'vue';
import MockDate from 'mockdate';
import Descriptions from '..';
import { resetWarned } from '../../_util/warning';
import { asyncExpect } from '../../../tests/utils';

describe('Descriptions', () => {
  const errorSpy = jest.spyOn(console, 'error').mockImplementation(() => {});

  afterEach(() => {
    MockDate.reset();
    errorSpy.mockReset();
  });

  afterAll(() => {
    errorSpy.mockRestore();
  });

  it('when typeof column is object', async () => {
    const wrapper = mount(
      {
        render() {
          return (
            <Descriptions ref="descriptions" column={{ xs: 8, sm: 16, md: 24 }}>
              <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
              <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
              <Descriptions.Item label="time">18:00:00</Descriptions.Item>
              <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
            </Descriptions>
          );
        },
      },
      { sync: false, attachTo: 'body' },
    );
    await asyncExpect(() => {
      expect(
        wrapper.findAll('td').reduce((total, td) => total + parseInt(td.attributes().colspan), 0),
      ).toBe(8);
    }, 100);
    wrapper.unmount();
  });

  it('column is number', () => {
    // eslint-disable-next-line global-require
    const wrapper = mount({
      render() {
        return (
          <Descriptions column={3}>
            <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
            <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
            <Descriptions.Item label="time">18:00:00</Descriptions.Item>
            <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
          </Descriptions>
        );
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    wrapper.unmount();
  });

  it('warning if ecceed the row span', () => {
    resetWarned();

    mount({
      render() {
        return (
          <Descriptions column={3}>
            <Descriptions.Item label="Product" span={2}>
              Cloud Database
            </Descriptions.Item>
            <Descriptions.Item label="Billing" span={2}>
              Prepaid
            </Descriptions.Item>
          </Descriptions>
        );
      },
    });
    expect(errorSpy).toHaveBeenCalledWith(
      'Warning: [ant-design-vue: Descriptions] Sum of column `span` in a line not match `column` of Descriptions.',
    );
  });

  it('when item is rendered conditionally', () => {
    const hasDiscount = false;
    const wrapper = mount({
      render() {
        return (
          <Descriptions>
            <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
            <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
            <Descriptions.Item label="time">18:00:00</Descriptions.Item>
            <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
            {hasDiscount && <Descriptions.Item label="Discount">$20.00</Descriptions.Item>}
          </Descriptions>
        );
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    wrapper.unmount();
  });

  it('vertical layout', () => {
    // eslint-disable-next-line global-require
    const wrapper = mount({
      render() {
        return (
          <Descriptions layout="vertical">
            <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
            <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
            <Descriptions.Item label="time">18:00:00</Descriptions.Item>
            <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
          </Descriptions>
        );
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
    wrapper.unmount();
  });

  it('Descriptions.Item support className', () => {
    const wrapper = mount({
      render() {
        return (
          <Descriptions>
            <Descriptions.Item label="Product" class="my-class">
              Cloud Database
            </Descriptions.Item>
          </Descriptions>
        );
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Descriptions support colon', () => {
    const wrapper = mount({
      render() {
        return (
          <Descriptions colon={false}>
            <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
          </Descriptions>
        );
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('Descriptions support style', () => {
    const wrapper = mount({
      render() {
        return (
          <Descriptions style={{ backgroundColor: '#e8e8e8' }}>
            <Descriptions.Item>Cloud Database</Descriptions.Item>
          </Descriptions>
        );
      },
    });
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('when max-width: 575pxï¼Œcolumn=1', async () => {
    // eslint-disable-next-line global-require
    const wrapper = mount(
      {
        render() {
          return (
            <Descriptions>
              <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
              <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
              <Descriptions.Item label="time">18:00:00</Descriptions.Item>
              <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
              <Descriptions.Item>No-Label</Descriptions.Item>
            </Descriptions>
          );
        },
      },
      { sync: false, attachTo: 'body' },
    );
    await asyncExpect(() => {
      expect(wrapper.findAll('tr')).toHaveLength(5);
      expect(wrapper.findAll('.ant-descriptions-item-label')).toHaveLength(4);
    });

    wrapper.unmount();
  });

  it('when max-width: 575pxï¼Œcolumn=2', async () => {
    const wrapper = mount(
      {
        render() {
          return (
            <Descriptions column={{ xs: 2 }}>
              <Descriptions.Item label="Product">Cloud Database</Descriptions.Item>
              <Descriptions.Item label="Billing">Prepaid</Descriptions.Item>
              <Descriptions.Item label="time">18:00:00</Descriptions.Item>
              <Descriptions.Item label="Amount">$80.00</Descriptions.Item>
            </Descriptions>
          );
        },
      },
      { sync: false, attachTo: 'body' },
    );
    await asyncExpect(() => {
      expect(wrapper.findAll('tr')).toHaveLength(2);
    });
    wrapper.unmount();
  });

  it('columns 5 with customize', () => {
    const wrapper = mount({
      render() {
        return (
          <Descriptions layout="vertical" column={4}>
            {/* 1 1 1 1 */}
            <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
            <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
            <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
            <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
            {/* 2 2 */}
            <Descriptions.Item label="bamboo" span={2}>
              bamboo
            </Descriptions.Item>
            <Descriptions.Item label="bamboo" span={2}>
              bamboo
            </Descriptions.Item>
            {/* 3 1 */}
            <Descriptions.Item label="bamboo" span={3}>
              bamboo
            </Descriptions.Item>
            <Descriptions.Item label="bamboo">bamboo</Descriptions.Item>
          </Descriptions>
        );
      },
    });

    function matchSpan(rowIndex, spans) {
      const tr = wrapper.findAll('tr')[rowIndex];
      const tds = tr.findAll('th');
      expect(tds.length).toEqual(spans.length);
      tds.forEach((td, index) => {
        expect(parseInt(td.attributes().colspan)).toEqual(spans[index]);
      });
    }

    matchSpan(0, [1, 1, 1, 1]);
    matchSpan(2, [2, 2]);
    matchSpan(4, [3, 1]);
  });

  it('number value should render correct', () => {
    const wrapper = mount({
      render() {
        return (
          <Descriptions bordered>
            <Descriptions.Item label={0}>{0}</Descriptions.Item>
          </Descriptions>
        );
      },
    });

    expect(wrapper.find('th').classes()).toContain('ant-descriptions-item-label');
    expect(wrapper.find('td').classes()).toContain('ant-descriptions-item-content');
  });

  it('Descriptions support extra', async () => {
    const wrapper = mount(Descriptions, {
      props: {
        extra: 'Edit',
      },
      slots: {
        default: h(
          Descriptions.Item,
          {
            label: 'UserName',
          },
          'Zhou Maomao',
        ),
      },
    });

    await asyncExpect(() => {
      expect(wrapper.find('.ant-descriptions-extra').exists()).toBe(true);
    });

    wrapper.setProps({ extra: undefined });

    await asyncExpect(() => {
      expect(wrapper.find('.ant-descriptions-extra').exists()).toBe(false);
    });
  });
});
