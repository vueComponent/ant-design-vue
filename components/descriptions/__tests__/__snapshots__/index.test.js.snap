// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Descriptions Descriptions support colon 1`] = `
<div class="ant-descriptions">
  <!---->
  <div class="ant-descriptions-view">
    <table>
      <tbody>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label ant-descriptions-item-no-colon">Product</span><span class="ant-descriptions-item-content">Cloud Database</span></div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions Descriptions support style 1`] = `
<div style="background-color: rgb(232, 232, 232);" class="ant-descriptions">
  <!---->
  <div class="ant-descriptions-view">
    <table>
      <tbody>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container">
              <!----><span class="ant-descriptions-item-content">Cloud Database</span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions Descriptions.Item support className 1`] = `
<div class="ant-descriptions">
  <!---->
  <div class="ant-descriptions-view">
    <table>
      <tbody>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item my-class" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">Product</span><span class="ant-descriptions-item-content">Cloud Database</span></div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions column is number 1`] = `
<div class="ant-descriptions">
  <!---->
  <div class="ant-descriptions-view">
    <table>
      <tbody>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">Product</span><span class="ant-descriptions-item-content">Cloud Database</span></div>
          </td>
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">Billing</span><span class="ant-descriptions-item-content">Prepaid</span></div>
          </td>
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">time</span><span class="ant-descriptions-item-content">18:00:00</span></div>
          </td>
        </tr>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item" colspan="3">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">Amount</span><span class="ant-descriptions-item-content">$80.00</span></div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions vertical layout 1`] = `
<div class="ant-descriptions">
  <!---->
  <div class="ant-descriptions-view">
    <table>
      <tbody>
        <tr class="ant-descriptions-row">
          <th class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">Product</span>
              <!---->
            </div>
          </th>
        </tr>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container">
              <!----><span class="ant-descriptions-item-content">Cloud Database</span>
            </div>
          </td>
        </tr>
        <tr class="ant-descriptions-row">
          <th class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">Billing</span>
              <!---->
            </div>
          </th>
        </tr>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container">
              <!----><span class="ant-descriptions-item-content">Prepaid</span>
            </div>
          </td>
        </tr>
        <tr class="ant-descriptions-row">
          <th class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">time</span>
              <!---->
            </div>
          </th>
        </tr>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container">
              <!----><span class="ant-descriptions-item-content">18:00:00</span>
            </div>
          </td>
        </tr>
        <tr class="ant-descriptions-row">
          <th class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">Amount</span>
              <!---->
            </div>
          </th>
        </tr>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container">
              <!----><span class="ant-descriptions-item-content">$80.00</span>
            </div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;

exports[`Descriptions when item is rendered conditionally 1`] = `
<div class="ant-descriptions">
  <!---->
  <div class="ant-descriptions-view">
    <table>
      <tbody>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">Product</span><span class="ant-descriptions-item-content">Cloud Database</span></div>
          </td>
        </tr>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">Billing</span><span class="ant-descriptions-item-content">Prepaid</span></div>
          </td>
        </tr>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">time</span><span class="ant-descriptions-item-content">18:00:00</span></div>
          </td>
        </tr>
        <tr class="ant-descriptions-row">
          <td class="ant-descriptions-item" colspan="1">
            <div class="ant-descriptions-item-container"><span class="ant-descriptions-item-label">Amount</span><span class="ant-descriptions-item-content">$80.00</span></div>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
